Remove iOS 13 deprecated APIs

ualch9 opened the pull request:
Allows OBAKit and OneBusAway to compile with a minimum iOS 13 target without Xcode complaining. 
~The project's minimum build target was not changed, it is still iOS 11.~
Edit 1/26/20: Project minimum build target is now iOS 11.4.

- Replace UITableViewRowAction with UIContextualAction
- Rows with a non-nil `deleteModel` will receive a FREE delete contextual swipe action, courtesy of OBAStaticTableViewController
- Update NSKeyedArchiver & NSKeyedUnarchiver API for iOS 13 (note below)
- Add "Radio Technology" & support for multiple carriers to email support HTML message
- Enable Large Content Viewer for buttons in OBAButtonBarRow (see WWDC 2019 session 261)
- Darkify warning icons (for service alerts)
- Update minimum build target of app and OBAKit to iOS 11.4

## NSKeyedArchiver/NSKeyedUnarchiver note
I am unfamiliar with NSCoding and NSSecureCoding so I didn't update NSCoding objects to NSSecureCoding, as suggested by Apple, for this PR. The API responsible for unsafely decoding root level objects in Objective-C was removed in iOS 12, so I had to bridge it from Swift manually (`OBAKit/Helpers/NSKeyedUnarchiver+UnarchiveTopLevelObjectObjC.swift`). Maybe I'll try NSSecureCoding at a later time.

ualch9 committed:
Darkify warning icons

aaronbrethorst commented on OBAKit/Helpers/NSKeyedUnarchiver+UnarchiveTopLevelObjectObjC.swift:
```diff
@@ -0,0 +1,21 @@
+import Foundation
+
+extension NSKeyedUnarchiver {
+    /// Obj-C compatible method of unarchiving NSCoding objects for iOS 13. Don't use this for NSSecureCoding objects.
```
Can you elaborate on why having two code paths—one for iOS 13 and one for iOS 11 and 12 is preferable to what exists today?

ualch9 replied:
updated code documentation to better explain this.

technically, because we still building against iOS 11, we can revert back to using the deprecated method. xcode won't complain about it unless we build against iOS 13.

aaronbrethorst commented on OBAKit/Location/OBARegionStorage.m:
```diff
@@ -84,14 +90,14 @@ - (void)writeRegionsToDisk:(NSArray<OBARegionV2*>*)regions {
+        DDLogError(@"Failed to unarchive persisted regions: %@", error);
     }
-
+    
```
make sure you enable the two "automatically trim trailing whitespace" options in Xcode's preferences. 

<img width="912" alt="image" src="https://user-images.githubusercontent.com/2254/73140267-f2d46a80-402b-11ea-9448-50e422b14c6f.png">

aaronbrethorst commented on OBAKit/OBAKit.xcodeproj/project.pbxproj:
```diff
@@ -1780,7 +1784,7 @@
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
-				IPHONEOS_DEPLOYMENT_TARGET = 11.0;
+				IPHONEOS_DEPLOYMENT_TARGET = 13.2;
```
what target is this for?

ualch9 replied:
oops, unintended change left over from testing

aaronbrethorst commented on OBAKit/UI/Controls/StackedButton.swift:
```diff
@@ -42,15 +42,28 @@ public class StackedButton: UIControl {
         accessibilityTraits = .button
         isUserInteractionEnabled = true
-
+        
```
whitespace

aaronbrethorst commented on OBAKit/UI/Controls/StackedButton.swift:
```diff
@@ -42,15 +42,28 @@ public class StackedButton: UIControl {
-
+        
 		if #available(iOS 13, *) {
+            showsLargeContentViewer = true
```
It looks like you have indentation with tabs instead of spaces turned on in Xcode.

<img width="912" alt="image" src="https://user-images.githubusercontent.com/2254/73140290-362ed900-402c-11ea-8cbd-48ea57f00da2.png">

aaronbrethorst commented on OBAKit/UI/MVVM/OBAStaticTableViewController.m:
```diff
@@ -208,11 +209,11 @@ - (BOOL)replaceRowAtIndexPath:(NSIndexPath*)indexPath withRow:(OBABaseRow*)row {
 
     OBAGuard(tableRow.deleteModel) else {
-        return;
+        return false;
```
Objective-C should use `YES` and `NO` https://nshipster.com/bool/

aaronbrethorst commented on OBAKit/UI/MVVM/OBAStaticTableViewController.m:
```diff
@@ -334,14 +336,31 @@ - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath
+            completionHandler(deleteStatus);
+        }];
+        
+        if (!actionsConfiguration) {
```
you don't need to change this, but I generally find it clearer to put the 'positive' case first. i.e. "if an `actionsConfiguration` object exists, then do this, otherwise do something else," instead of "if an `actionsConfiguration` object DOES NOT exist..."

When it comes to readability, there's a lot riding on an easily-missed `!`

ualch9 replied:
I agree with you, updated.

aaronbrethorst commented on OBAKit/UI/MVVM/OBAStaticTableViewController.m:
```diff
@@ -334,14 +336,31 @@ - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath
+        
+        return configurationWithDeleteAction;
+    } else {
+        return row.rowActions;
```
In contrast with what I wrote above, I prefer early returns over multiple layers of `if` statement nesting, even when that requires a negation. That way, instead of two levels of `if` statement nesting in this method, you can have a few single layers with an early return at the beginning of the method. i.e.

    if (!row.deleteModel) {
        return row.rowActions;
    }

ualch9 replied:
agreed.

aaronbrethorst commented on OneBusAway/OneBusAwayTests/OBAKitTests/models/OBAAgencyV2_Tests.m:
```diff
@@ -40,9 +40,9 @@ - (void)testCoding {
-
-    OBAAgencyV2 *unarchived = [NSKeyedUnarchiver unarchiveObjectWithData:archivedData];
+    NSData *archivedData = [NSKeyedArchiver archivedDataWithRootObject:agency requiringSecureCoding:false error:nil];
+    
```
whitespace

aaronbrethorst commented on OneBusAway/ui/regions/RegionListViewController.swift:
```diff
@@ -109,29 +109,23 @@ extension RegionListViewController {
         }
-
-        let delete = UITableViewRowAction.init(style: .destructive, title: OBAStrings.delete) { (_, indexPath) in
+        
```
whitespace

aaronbrethorst commented on OneBusAway/ui/search/MapSearchViewController.swift:
```diff
@@ -185,6 +185,11 @@ class MapSearchViewController: OBAStaticTableViewController, UISearchResultsUpda
         let tabBarHeight: CGFloat = 49
         self.tableView.contentInset.bottom = tabBarHeight
-        self.tableView.scrollIndicatorInsets.bottom = tabBarHeight
+        if #available(iOS 11.1, *) {
```
I'd rather see the deployment target get updated to 11.4 than an availability check for a minor version of iOS that came out in 2017.

aaronbrethorst reviewed (changes requested):
Generally, this looks very good! I have just a couple questions and some nits.

aaronbrethorst commented on OBAKit/Location/OBARegionStorage.m:
```diff
@@ -84,12 +90,12 @@ - (void)writeRegionsToDisk:(NSArray<OBARegionV2*>*)regions {
 
-    NSArray<OBARegionV2*>* regions = nil;
+    NSData *objectData = [NSData dataWithContentsOfFile:OBARegionStorage.localRegionsFilePath];
+    NSError *error;
```
Sorry I didn't catch this before, but many APIs will fill their error parameters with garbage when the API returns successfully. A safer check is:

    NSError *error = nil;
    NSArray *regions = [NSKeyedUnarchiver unarchiveTopLevelObjectFor:objectData error:&error];
    if (!regions) {
         DDLogError(@"Failed to unarchive persisted regions: %@", error);
     }

aaronbrethorst commented on OneBusAway/ui/map/OBAMapViewController.m:
```diff
@@ -660,14 +660,26 @@ - (void)setMostRecentRegion:(MKCoordinateRegion)mostRecentRegion {
     NSData *data = [userDefaults objectForKey:@"mostRecentRegion"];
-    OBABoxedCoordinateRegion *boxedRegion = [NSKeyedUnarchiver unarchiveObjectWithData:data];
+    NSError *error;
+    OBABoxedCoordinateRegion *boxedRegion = [NSKeyedUnarchiver unarchiveTopLevelObjectFor:data error:&error];
```
same here as in the last comment

aaronbrethorst commented:
looking good! I noticed one potential issue if Apple's API returns garbage to us, which I noted in a pair of comments. I'll merge as soon as that's addressed.

ualch9 committed:
Remove iOS 13 deprecated APIs

- Update minimum build target to iOS 11.4.
- Move responsibility of adding a delete swipe action in OBAStaticTableViewController from manually adding to row.rowActions to automatically including a delete action for rows with a non-nil deleteModel.
- Update NSKeyedArchiver and NSKeyedUnarchiver API for iOS 13.
- Add "Radio Technology" and support for multiple carriers to email support HTML message.
- Large content mode for actions on buttons in OBAButtonBarRow (see WWDC 2019 session 261).

aaronbrethorst reviewed (approved):
👍 nice work
